### PR TITLE
WIP unit tests for Router

### PR DESCRIPTION
**Pull request by brainthinks**
a few unit tests for Router, which test the "entry point", as well as the "onunload" function, using mocks.

### CODE REVIEW

File: src/js/conduit/Router.js [hunk 1]
```diff
@@ -839,6 +840,7 @@ export default function () {
     },
     onunload: function () {
       if (!window.router) {
+        // window.router.logger.info('onunload - Router not instantiated, exiting gracefully');
```
Comment (csmcgee) [inline src/js/conduit/Router.js]: debug code?

File: src/js/conduit/Router.test.js [hunk 2]
```diff
@@ -2,6 +2,124 @@ import '@babel/polyfill';
 
 import Router from './Router';
 
-test('adds 1 + 2 to equal 3', async () => {
-  expect(1 + 2).toBe(3);
+describe('Router', () => {
+  describe('export value', () => {
```
Comment (csmcgee) [inline src/js/conduit/Router.test.js]: @brainthinks review naming conventions for tests - tests can be more english than robot.

File: src/js/conduit/Router.test.js [hunk 3]
```diff
@@ -2,6 +2,124 @@ import '@babel/polyfill';
...
+    test('is function', () => {
+      expect(typeof Router).toBe('function');
+    });
+  });
+
+  describe('has everything necessary for iframe functionality', () => {
```
Comment (csmcgee) [inline src/js/conduit/Router.test.js]: @brainthinks we should review the pros and cons of property testing. I don't recall them off hand but basically some consider them pretty low value. but maybe it's okay.

Also.... could be other test tools for property testing and/or extensions to make it easier.

File: src/js/conduit/Router.test.js [hunk 4]
```diff
@@ -2,6 +2,124 @@ import '@babel/polyfill';
...
+        const {
+          onunload,
+        } = Router();
+
+        window.router = generateMockRouter();
+
+        onunload();
+
+        expect(window.router.destroy.mock.calls.length).toBe(1);
+        expect(window.router.logger.info.mock.calls.length).toBe(2);
```
Comment (csmcgee) [inline src/js/conduit/Router.test.js]: @brainthinks the logging checks seem pretty weak to me - I know we do it inside of the kong tests but that was left over from previous developers. Something to review as well.

File: src/js/conduit/Router.test.js [hunk 5]
```diff
@@ -2,6 +2,124 @@ import '@babel/polyfill';
...
+
+        window.router.destroy = jest.fn(() => {
+          throw new Error('Unspecified Error');
+        });
+
+        onunload();
+
+        expect(window.router.destroy.mock.calls.length).toBe(1);
+        expect(window.router.logger.info.mock.calls.length).toBe(1);
+        expect(window.router.logger.error.mock.calls.length).toBe(1);
```
Comment (csmcgee) [inline src/js/conduit/Router.test.js]: @brainthinks being dependent on the logger for these tests seem lame..... I don't have a solution right now though.

File: src/js/conduit/Router.test.js (hunk 2 above)
Comment (csmcgee) [reply src/js/conduit/Router.test.js]: see https://github.com/facebook/react/blob/f625fce8578cc77113d9da8ffd02032a60477e6a/packages/react-dom/src/events/__tests__/SyntheticKeyboardEvent-test.js#L35 for example

File: src/js/conduit/Router.test.js [hunk 6]
```diff
@@ -2,6 +2,124 @@ import '@babel/polyfill';
...
+    });
+  });
+
+  describe('has everything necessary for iframe functionality', () => {
+    test('returns object', () => {
+      const router = Router();
+
+      expect(typeof router).toBe('object');
+    });
+    test('has "Router" property, which is the Router class', () => {
```
Comment (csmcgee) [inline src/js/conduit/Router.test.js]: this is pretty descriptive maybe even change `"Router"` to `.Router` to be close to code.

File: src/js/conduit/Router.test.js (hunk 3 above)
Comment (csmcgee) [reply src/js/conduit/Router.test.js]: check out https://github.com/pact-foundation/pact-js or other js property based testing frameworks.

**Comment by csmcgee** [comment]
see https://github.com/goldbergyoni/javascript-testing-best-practices#-%EF%B8%8F-12-structure-tests-by-the-aaa-pattern for structuring your tests. Also a good resource for testing in general

File: src/js/conduit/Router.js (hunk 1 above)
Comment (brainthinks) [reply src/js/conduit/Router.js]: done

File: src/js/conduit/Router.test.js (hunk 2 above)
Comment (brainthinks) [reply src/js/conduit/Router.test.js]: i have re-written all of the "describe" and "it" strings

File: src/js/conduit/Router.test.js (hunk 3 above)
Comment (brainthinks) [reply src/js/conduit/Router.test.js]: pact appears to be focused on api testing.  I ended up using jest-extended for additional object assertions

File: src/js/conduit/Router.test.js (hunk 4 above)
Comment (brainthinks) [reply src/js/conduit/Router.test.js]: done

File: src/js/conduit/Router.test.js (hunk 5 above)
Comment (brainthinks) [reply src/js/conduit/Router.test.js]: done

File: src/js/conduit/Router.test.js (hunk 6 above)
Comment (brainthinks) [reply src/js/conduit/Router.test.js]: I ended up consolidating the individual property assertions into a single test

**Comment by brainthinks** [comment]
I also added eslint.  this required certain rules / configs to be defined, so I structured things similarly to react

**Comment by brainthinks** [comment]
Closing because I implemented these changes in STRY0060808

**Pull request closed without merging**